examples/async-recursion/src/lib.rs

```diff
@@ -0,0 +1,44 @@
+#![no_std]
+
+use core::num::ParseIntError;
+use gstd::{msg, prelude::*, ActorId};
+use async_recursion::async_recursion;
+
+static mut DEST: ActorId = ActorId::new([0u8; 32]);
+
+fn decode_hex(s: &str) -> Result<Vec<u8>, ParseIntError> {
+    (0..s.len())
+        .step_by(2)
+        .map(|i| u8::from_str_radix(&s[i..i + 2], 16))
+        .collect()
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
+    let input = String::from_utf8(msg::load_bytes()).expect("Invalid message: should be utf-8");
+    DEST = ActorId::from_slice(
+        &decode_hex(&input).expect("INTIALIZATION FAILED: INVALID PROGRAM ID"),
+    )
+    .expect("Unable to create ActorId");
+}
+
+/// Send message "PING" and waits for reply, then recursevelly
+/// repeats with `val` decreased by reply len, until val > 0.
+#[async_recursion]
+async fn rec_func(val: u32) {
+    let reply = msg::send_bytes_and_wait_for_reply(unsafe { DEST }, b"PING", 0)
+        .expect("Error in sending message")
+        .await
+        .expect("Error in async message processing");
+
+    msg::send_bytes(msg::source(), format!("Hello, val = {}", val), 0).unwrap();
+
+    if val - reply.len() as u32 > 0 {
+        rec_func(val - reply.len() as u32).await;
```

Comment:
Another tiny remark is that `value` could have `usize` type so there is no need to cast it to u32:
```rust
if val > reply.len() {
    rec_func(val - reply.len()).await;
```

Reply:
Done